### PR TITLE
[Common] Skip cuDNN 9.10.0/9.10.1 due to bugs

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

cuDNN fixed a few bugs in cuDNN 9.10.2 and this PR is to:
- skip cuDNN 9.10.0 for SDPA FP8 related bugs,
- skip cuDNN 9.10.0 and 9.10.1 for SDPA FP16/BF16 related bugs,
- unify the use of `ModelConfig` in all unit tests,
- fix FP8 and certain FP16/BF16 tests so they check backend availability first before running the configs,
- move `test_sanity_attention_extra_state` from `test_sanity.py` to `test_fused_attn.py`,
- remove `test_kv_cache_accuracy` from `test_numerics.py` because it's duplicate to `test_kv_cache` in `test_kv_cache.py`, and
- rename `tests/pytorch/fused_attn` to `tests/pytorch/attention` and other relevant files from `fused_attn` to `attention`.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please see description.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by cyanguwa** [comment]
Pipeline 31306522 for 9.10.0 + 25.06 + 12.9
Pipeline 31306722 for 9.10.1 + 25.06 + 12.9

File: tests/pytorch/test_sanity.py [hunk 1]
```diff
@@ -47,6 +47,7 @@
 from transformer_engine.pytorch.tensor.utils import replace_raw_data
 from transformer_engine.pytorch.distributed import checkpoint
 from test_numerics import reset_rng_states, dtype_tols
+from fused_attn.test_fused_attn import ModelConfig, _get_attention_backends
```
Comment (timmoon10) [inline tests/pytorch/test_sanity.py]: `test_fused_attn.py` is kind of a strange place to keep `ModelConfig`. It's indicating that these tests are becoming over-tuned to attention. We could avoid this refactor by moving `test_sanity_attention_extra_state` into the fused attention tests:
https://github.com/NVIDIA/TransformerEngine/blob/7b249ae82d92432936619fbe51f119c5f148a89c/tests/pytorch/test_sanity.py#L1119

Alternatively, we could move `ModelConfig` from `test_fused_attn.py` into `utils.py`. It has many attention-specific options though, so I'd prefer having separate implementations.

**Review by zhongbozhu — DISMISSED** [review]
LGTM

File: tests/pytorch/test_sanity.py (hunk 1 above)
Comment (cyanguwa) [reply tests/pytorch/test_sanity.py]: I've moved `test_sanity_attention_extra_state` to `test_fused_attn.py`, which is now renamed to `test_attention.py`. I've also moved `ModelConfig`, and a couple of other utility functions, to `tests/pytorch/utils.py`. Please take a look and let me know if the structure is all right! `ModelConfig` has a lot of parameters which seem to be closely related to attention but I think we need them for other tests as well, like `hidden_size`, `head_dim`, and `batch_size`. I plumbed it through and I don't think it's too clunky. Anyway, let me know!

File: tests/pytorch/utils.py [hunk 2]
```diff
@@ -4,12 +4,31 @@
...
+    AttentionLogging,
+)
+from transformer_engine.pytorch.cpp_extensions.fused_attn import FusedAttnBackend
+
+# Initialize RNG state
+seed = 1234
+torch.manual_seed(seed)
+torch.cuda.manual_seed(seed)
+_cpu_rng_state = torch.get_rng_state()
+_cuda_rng_state = torch.cuda.get_rng_state()
```
Comment (timmoon10) [inline tests/pytorch/utils.py]: Importing a module shouldn't have side effects. How about something like:
```python
# Cached RNG state
_rng_state: Optional[Tuple[torch.Tensor, torch.Tensor]] = None

def reset_rng_state() -> None:
    """Revert to deterministic RNG state"""
    global _rng_state
    if _rng_state is None:
        torch.manual_seed(1234)
        torch.cuda.manual_seed(1234)
        _rng_state = (torch.get_rng_state(), torch.cuda.get_rng_state())
    else:
        cpu_rng_state, cuda_rng_state = _rng_state
        torch.set_rng_state(cpu_rng_state)
        torch.cuda.set_rng_state(cuda_rng_state)
```
Each test would need to call `reset_rng_state` explicitly if it wanted deterministic RNG state.

File: tests/pytorch/test_sanity.py (hunk 1 above)
Comment (timmoon10) [reply tests/pytorch/test_sanity.py]: There are many attention-specific args in `ModelConfig`. For linear layers, the only configs that matter are the BSHD dims so we can initialize the data tensors. I still think it's better to have a basic config in `test_numerics.py` and an attention-specific config in `test_attention.py`. At the least, `ModelConfig` should have defaults so `test_numerics.py` doesn't need much attention-specific knowledge.

File: tests/pytorch/test_numerics.py [hunk 3]
```diff
@@ -124,6 +114,18 @@ def __init__(self, hidden_size, eps, num_attention_heads, embed, num_layers, seq
...
+def is_fused_attn_available(
+    config: ModelConfig, dtype: torch.dtype, qkv_layout="bshd_bshd_bshd", is_training=True
+):
+    available_backends, _, fused_attn_backends = get_available_attention_backends(
+        config,
+        qkv_dtype=dtype,
+        qkv_layout=qkv_layout,
+        is_training=is_training,
+    )
+    return FusedAttnBackend["F16_arbitrary_seqlen"] in fused_attn_backends
```
Comment (timmoon10) [inline tests/pytorch/test_numerics.py:123]: I'm uncomfortable that tests in `test_numerics.py` require knowledge of attention backends and would prefer if they were in `test_attention.py` instead. However, I see that this is needed for some of the full GPT model tests, which do belong in this file.

**Review by timmoon10 — COMMENTED** [review]
Perhaps we should split this PR into two: skipping the buggy cuDNN versions and refactoring the attention tests.

File: tests/pytorch/utils.py (hunk 2 above)
Comment (cyanguwa) [reply tests/pytorch/utils.py]: I incorporated this in the code. Thanks!

File: tests/pytorch/test_numerics.py (hunk 3 above)
Comment (cyanguwa) [reply tests/pytorch/test_numerics.py:123]: Any test that's based on `TransformerLayer` will need some knowledge of attention, and there are a lot of those tests even outside `tests/pytorch/attention`, like `test_numerics.py`, `test_cpu_offload.py`, `test_cuda_graphs.py` and `test_sanity.py`. I feel it's difficult to pull them apart - at the end of the day, we are working with LLMs. I've simplified `ModelConfig` a bit though, so for some of the tests, users only need to specify ~5 params to instantiate a `ModelConfig`.

File: tests/pytorch/test_sanity.py (hunk 1 above)
Comment (cyanguwa) [reply tests/pytorch/test_sanity.py]: Please see comment below.

File: tests/pytorch/utils.py [hunk 4]
```diff
@@ -4,12 +4,29 @@
...
+    get_attention_backend,
+    AttentionParams,
+    AttentionLogging,
+)
+from transformer_engine.pytorch.cpp_extensions.fused_attn import FusedAttnBackend
+
+# Initialize RNG state
+seed = 1234
+torch.manual_seed(seed)
+torch.cuda.manual_seed(seed)
```
Comment (timmoon10) [inline tests/pytorch/utils.py]: Importing a module shouldn't affect the global RNG state, since that could cause the RNG to change at unexpected times.
```suggestion
```

File: tests/pytorch/test_numerics.py [hunk 5]
```diff
@@ -59,30 +61,18 @@
 torch.manual_seed(seed)
 torch.cuda.manual_seed(seed)
 # Record initial RNG state from script run.
-_cpu_rng_state = torch.get_rng_state()
-_cuda_rng_state = torch.cuda.get_rng_state()
+_rng_states = None
```
Comment (timmoon10) [inline tests/pytorch/test_numerics.py]: Setting the RNG state is redundant with `reset_rng_states`. 
```suggestion
```
There are a few other files where we redundantly set the RNG state.

File: tests/pytorch/utils.py [hunk 6]
```diff
@@ -106,3 +123,184 @@ def make_recipe(name: Optional[str]) -> Optional[Recipe]:
...
+
+class ModelConfig:
+    def __init__(
+        self,
+        batch_size: int,
+        num_heads: int,
+        num_gqa_groups: int,
+        head_dim_qk: int,
+        max_seqlen_q: int,
+        max_seqlen_kv: int,
```
Comment (timmoon10) [inline tests/pytorch/utils.py]: Nit: I think `num_gqa_groups=num_heads` and `max_seqlen_kv=max_seqlen_q` are reasonable defaults. This way we have the minimum number of args needed for a basic Transformer layer.
```suggestion
    def __init__(
        self,
        batch_size: int,
        num_heads: int,
        head_dim_qk: int,
        max_seqlen_q: int,
        num_gqa_groups: Optional[int] = None,
        max_seqlen_kv: Optional[int] = None,
```

File: tests/pytorch/test_numerics.py (hunk 3 above)
Comment (timmoon10) [reply tests/pytorch/test_numerics.py:123]: Having good defaults in `ModelConfig` definitely helps a lot. I don't mind having attention-specific stuff internally, as long as it doesn't affect anything non-attention-related.

File: tests/pytorch/test_numerics.py [hunk 7]
```diff
@@ -40,11 +40,13 @@
 from transformer_engine.pytorch.attention.inference import InferenceParams
 from transformer_engine.pytorch.distributed import checkpoint as te_checkpoint
 from transformer_engine.pytorch.cpp_extensions import general_gemm, general_grouped_gemm
+from transformer_engine.pytorch.cpp_extensions.fused_attn import FusedAttnBackend
 from transformer_engine.pytorch.tensor.float8_tensor import Float8Quantizer
 from transformer_engine.pytorch.module.base import get_multi_stream_cublas_workspace, get_workspace
 from transformer_engine.pytorch.utils import get_device_compute_capability, get_cudnn_version
 from transformer_engine.common import recipe
 import transformer_engine_torch as tex
+from utils import ModelConfig, _rng_states, reset_rng_states, get_available_attention_backends
```
Comment (timmoon10) [inline tests/pytorch/test_numerics.py]: No need to import `_rng_states` unless we want to do something with it. It's an implementation detail of `reset_rng_states`, so importing it should be viewed as an evil hack.
```suggestion
from utils import ModelConfig, reset_rng_states, get_available_attention_backends
```
We import it in a few other files.

File: tests/pytorch/utils.py (hunk 6 above)
Comment (cyanguwa) [reply tests/pytorch/utils.py]: I've shuffled the params a bit and made some kwargs.

**Review by timmoon10 — APPROVED** [review]
LGTM

**Review by timmoon10 — APPROVED** [review]

File: tests/pytorch/test_sanity.py [hunk 8]
```diff
@@ -105,37 +103,19 @@ def custom_amax_compute(amax_history: torch.Tensor) -> torch.Tensor:
...
-    kv_channels: Optional[int] = None
-
-    def is_fp8_supported(self):
-        if self.seq_len * self.batch_size % 16:
-            return False
-        if self.hidden_size % 16:
-            return False
-        return True
+def is_fp8_supported(config: ModelConfig):
+    if config.max_seqlen_q * config.batch_size % 16 or config.max_seqlen_q * config.batch_size % 16:
```
Comment (timmoon10) [inline tests/pytorch/test_sanity.py]:
```suggestion
    if config.max_seqlen_q * config.batch_size % 16 or config.max_seqlen_kv * config.batch_size % 16:
```

File: tests/pytorch/utils.py [hunk 9]
```diff
@@ -106,3 +118,184 @@ def make_recipe(name: Optional[str]) -> Optional[Recipe]:
...
+    else:
+        cpu_rng_state, cuda_rng_state = _rng_states
+        torch.set_rng_state(cpu_rng_state)
+        torch.cuda.set_rng_state(cuda_rng_state)
+
+
+@pytest.fixture(autouse=True)
+def reset_global_fp8_state():
+    yield
+    fp8.FP8GlobalStateManager.reset()
```
Comment (timmoon10) [inline tests/pytorch/utils.py]: Are we trying to apply this fixture in every test that includes `utils.py`? This is unintuitive and may break some tests. Also, I'm not sure if PyTest is smart enough to detect fixture registration from imported modules.
```suggestion
```
For now, it would probably be easiest to duplicate the code in each test like before.

Comment (cyanguwa) [reply tests/pytorch/utils.py]: Apparently, fixture import is not recommended. If an import happens, the fixture will be registered as defined in that particular module instead of shared by the importing module as well.
